File: tests/search/post_searcher/post_searcher.rb
# Copyright Vespa.ai. All rights reserved.
require 'indexed_streaming_search_test'

class Post_Searcher < IndexedStreamingSearchTest

  def setup
    set_owner("bjorncs")
    set_description("Test that we can POST to a searcher")
    add_bundle(selfdir + "PostSearcher.java")
    search_chain = SearchChain.new.
      add(Searcher.new("com.yahoo.example.PostSearcher"))
    deploy_app(SearchApp.new.sd(SEARCH_DATA+"music.sd").search_chain(search_chain))
    start
  end

  def test_twophase_searcher
    result = vespa.container.values.first.post_search("/search/?query=", "Hello world", 0, {'Content-Type' => 'text/plain'})
    assert(result.to_s =~ /Hello world/)
  end

  def teardown
    stop
  end

end

File: tests/search/post_searcher/PostSearcher.java
// Copyright Vespa.ai. All rights reserved.
package com.yahoo.example;

import com.yahoo.search.Query;
import com.yahoo.search.Result;
import com.yahoo.search.Searcher;
import com.yahoo.search.result.Hit;
import com.yahoo.search.searchchain.Execution;

import java.util.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class PostSearcher extends Searcher {

	@Override
    public Result search(Query query, Execution execution) {
		try {
			query.trace("running PostSearcher", false, 1);

			BufferedReader reader = new BufferedReader(new InputStreamReader(query.getHttpRequest().getData(), "UTF-8"));

			Result result = new Result(query);

			Hit h = new Hit("foo");
			h.setField("firstline", reader.readLine());
			result.hits().add(h);

			return result;
		} catch (Exception e) {
			return null;
		}
    }

}
